src/EditorFeatures/Core/Implementation/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+            {
+                InsertText(textChanges, selectedSpan.End, commentInfo.BlockCommentEndString);
+                InsertText(textChanges, selectedSpan.End, commentInfo.BlockCommentStartString);
+                // Shrink the tracking span so the next comment start marker is not included in selection.
+                amountToAddToEnd = -commentInfo.BlockCommentStartString.Length;
+            }
+
+            trackingSpans.Add(new CommentTrackingSpan(selectedSpan, amountToAddToStart, amountToAddToEnd));
+        }
+
+        private static void AddBlockComment(CommentSelectionInfo commentInfo, TextSpan span, List<TextChange> textChanges)
+        {
+            InsertText(textChanges, span.Start, commentInfo.BlockCommentStartString);
+            InsertText(textChanges, span.End, commentInfo.BlockCommentEndString);
+        }
+
+        private static void DeleteBlockComment(BlockCommentSelectionHelper blockCommentSelection, TextSpan spanToRemove,
+            List<TextChange> textChanges, CommentSelectionInfo commentInfo)
+        {
+            DeleteText(textChanges, new TextSpan(spanToRemove.Start, commentInfo.BlockCommentStartString.Length));
+            var endMarkerPosition = spanToRemove.End - commentInfo.BlockCommentEndString.Length;
+            // Sometimes the block comment will be missing a close marker.
+            if (Equals(blockCommentSelection.GetSubstringFromText(endMarkerPosition, commentInfo.BlockCommentEndString.Length),
+                commentInfo.BlockCommentEndString))
+            {
+                DeleteText(textChanges, new TextSpan(endMarkerPosition, commentInfo.BlockCommentEndString.Length));
+            }
+        }
+
+        private class BlockCommentSelectionHelper
+        {
+            private readonly string _text;
+            private readonly ITextSnapshot _snapshot;
+
+            public TextSpan SelectedSpan { get; }
+
+            public IEnumerable<TextSpan> IntersectingBlockComments { get; }
+
+            public IEnumerable<TextSpan> UncommentedSpansInSelection { get; }
+
+            public BlockCommentSelectionHelper(IEnumerable<TextSpan> allBlockComments, SnapshotSpan selectedSnapshotSpan)
+            {
+                _text = selectedSnapshotSpan.GetText().Trim();
+                _snapshot = selectedSnapshotSpan.Snapshot;
+
+                SelectedSpan = TextSpan.FromBounds(selectedSnapshotSpan.Start, selectedSnapshotSpan.End);
+                IntersectingBlockComments = GetIntersectingBlockComments(allBlockComments, SelectedSpan);
+                UncommentedSpansInSelection = GetUncommentedSpansInSelection();
+            }
+
+            /// <summary>
+            /// Determines if the given span is entirely whitespace.
+            /// </summary>
+            public bool IsSpanWhitespace(TextSpan span)
+            {
+                for (var i = span.Start; i < span.End; i++)
+                {
+                    if (!IsLocationWhitespace(i))
```

Comment:
i think you can just inline this method entirely.  it should just be `!char.IsWhitespcae(_snapshot[i])` right?